File: Sources/Xarita/UI/PaperBackground.swift
import SwiftUI

/// The drafting grid the whole window sits on.
///
/// A flat fill reads as an empty container; a ruled ground reads as a working
/// surface, which is the difference between the app feeling like a dialog and
/// feeling like something you work on. Drawn rather than tiled from an image so
/// it stays crisp at any scale factor and follows the theme.
struct PaperBackground: View {
    var spacing: CGFloat = 26

    var body: some View {
        Canvas { context, size in
            var path = Path()
            var x: CGFloat = 0
            while x <= size.width {
                path.move(to: CGPoint(x: x, y: 0))
                path.addLine(to: CGPoint(x: x, y: size.height))
                x += spacing
            }
            var y: CGFloat = 0
            while y <= size.height {
                path.move(to: CGPoint(x: 0, y: y))
                path.addLine(to: CGPoint(x: size.width, y: y))
                y += spacing
            }
            context.stroke(path, with: .color(Theme.paperGrid), lineWidth: 1)
        }
        .background(Theme.background)
        .drawingGroup()
        .allowsHitTesting(false)
    }
}

/// The three-bar terrain mark used for difficulty.
///
/// A number ("complexity 7") means nothing without a scale to compare it to.
/// Rising bars are read instantly and need no legend.
struct TerrainMark: View {
    let difficulty: GraphNode.Difficulty
    var scale: CGFloat = 1

    var body: some View {
        HStack(alignment: .bottom, spacing: 1.5 * scale) {
            ForEach(0..<3, id: \.self) { step in
                RoundedRectangle(cornerRadius: 0.5)
                    .fill(step <= difficulty.rawValue ? Theme.color(for: difficulty) : Theme.border)
                    .frame(width: 3 * scale, height: (5 + CGFloat(step) * 3) * scale)
            }
        }
    }
}

File: Windows/Atlas.Windows/Views/PaperBackground.cs
using Avalonia;
using Avalonia.Controls;
using Avalonia.Media;

namespace Atlas.Windows.Views;

/// <summary>
/// The drafting grid the whole window sits on.
///
/// A flat fill reads as an empty container; a ruled ground reads as a working
/// surface, which is the difference between the app feeling like a dialog and
/// feeling like something you work on. Drawn rather than tiled from an image
/// so it stays crisp at any scale factor and follows the theme.
///
/// Ported from Sources/Atlas/UI/PaperBackground.swift. SwiftUI's
/// <c>Canvas { context, size in … }</c> and Avalonia's
/// <c>Render(DrawingContext)</c> are the same idea — immediate mode over a
/// retained-mode toolkit — which is why the body below is line for line the
/// same drawing.
/// </summary>
public sealed class PaperBackground : Control
{
    public static readonly StyledProperty<double> SpacingProperty =
        AvaloniaProperty.Register<PaperBackground, double>(nameof(Spacing), 26);

    public double Spacing
    {
        get => GetValue(SpacingProperty);
        set => SetValue(SpacingProperty, value);
    }

    static PaperBackground()
    {
        AffectsRender<PaperBackground>(SpacingProperty);
        // Nothing here is interactive; the grid must never eat a click meant
        // for what is drawn on top of it.
        IsHitTestVisibleProperty.OverrideDefaultValue<PaperBackground>(false);
    }

    public override void Render(DrawingContext context)
    {
        var size = Bounds.Size;
        if (size.Width <= 0 || size.Height <= 0) return;

        context.FillRectangle(Broadsheet.Brush(Broadsheet.Background),
                              new Rect(0, 0, size.Width, size.Height));

        var pen = new Pen(Broadsheet.Brush(Broadsheet.PaperGrid), 1);
        for (double x = 0; x <= size.Width; x += Spacing)
            context.DrawLine(pen, new Point(x, 0), new Point(x, size.Height));
        for (double y = 0; y <= size.Height; y += Spacing)
            context.DrawLine(pen, new Point(0, y), new Point(size.Width, y));
    }
}
